exchanges/kline/kline.go

```diff
@@ -278,123 +264,91 @@ func durationToWord(in Interval) string {
-func (i *Interval) IntervalsPerYear() float64 {
-	if i.Duration() == 0 {
+func (i Interval) IntervalsPerYear() float64 {
+	if i == 0 {
 		return 0
 	}
-	return oneYearDurationInNano / float64(i.Duration().Nanoseconds())
+	return oneYearDurationInNano / float64(i)
 }
 
 // ConvertToNewInterval allows the scaling of candles to larger candles
-// eg convert OneDay candles to ThreeDay candles, if there are adequate candles
-// incomplete candles are NOT converted
-// eg an 4 OneDay candles will convert to one ThreeDay candle, skipping the fourth
-func ConvertToNewInterval(item *Item, newInterval Interval) (*Item, error) {
-	if item == nil {
+// e.g. Convert OneDay candles to ThreeDay candles, if there are adequate
+// candles. Incomplete candles are NOT converted e.g. 4 OneDay candles will
+// convert to one ThreeDay candle, skipping the fourth.
+func (k *Item) ConvertToNewInterval(newInterval Interval) (*Item, error) {
+	if k == nil {
 		return nil, errNilKline
 	}
+	if k.Interval <= 0 {
+		return nil, fmt.Errorf("%w for old candle", ErrInvalidInterval)
+	}
 	if newInterval <= 0 {
-		return nil, ErrUnsetInterval
+		return nil, fmt.Errorf("%w for new candle", ErrInvalidInterval)
 	}
-	if newInterval.Duration() <= item.Interval.Duration() {
-		return nil, ErrCanOnlyDownscaleCandles
+	if newInterval <= k.Interval {
+		return nil, fmt.Errorf("%w %s is less than or equal to %s",
+			ErrCanOnlyUpscaleCandles,
+			newInterval,
+			k.Interval)
 	}
-	if newInterval.Duration()%item.Interval.Duration() != 0 {
-		return nil, ErrWholeNumberScaling
+	if newInterval%k.Interval != 0 {
+		return nil, fmt.Errorf("%s %w %s",
+			k.Interval,
+			ErrWholeNumberScaling,
+			newInterval)
 	}
 
-	oldIntervalsPerNewCandle := int64(newInterval / item.Interval)
-	var candleBundles [][]Candle
-	candles := make([]Candle, 0, oldIntervalsPerNewCandle)
-	for i := range item.Candles {
-		candles = append(candles, item.Candles[i])
-		intervalCount := int64(i + 1)
-		if oldIntervalsPerNewCandle == intervalCount {
-			candleBundles = append(candleBundles, candles)
-			candles = candles[:0]
-		}
+	oldIntervalsPerNewCandle := int(newInterval / k.Interval)
+	candles := make([]Candle, len(k.Candles)/oldIntervalsPerNewCandle)
+	if len(candles) == 0 {
+		return nil, fmt.Errorf("%w to %v no candle data", ErrInsufficientCandleData, newInterval)
 	}
-	responseCandle := &Item{
-		Exchange: item.Exchange,
-		Pair:     item.Pair,
-		Asset:    item.Asset,
-		Interval: newInterval,
-	}
-	for i := range candleBundles {
-		var lowest, highest, volume float64
-		lowest = candleBundles[i][0].Low
-		highest = candleBundles[i][0].High
-		for j := range candleBundles[i] {
-			volume += candleBundles[i][j].Volume
-			if candleBundles[i][j].Low < lowest {
-				lowest = candleBundles[i][j].Low
-			}
-			if candleBundles[i][j].High > highest {
-				lowest = candleBundles[i][j].High
+	var target int
+	for x := range k.Candles {
+		if candles[target].Time.IsZero() {
+			candles[target].Time = k.Candles[x].Time
```

Comment:
This logic only accounts for perfect responses. Please see the following BTC Markets candle data response:
![image](https://user-images.githubusercontent.com/9261323/209271856-39a367e9-3fd6-450d-b117-e05933073828.png)

When data is missing, things get even more wacky as you assume the next time is time+interval. So all data beyond any missing data is incorrect

The function `FillMissingDataWithEmptyEntries` alleviates this for other purposes - but perhaps you can modify it to work for this scenario too

I would say you need to add in any missing candles when doing this, then add to the `kline.Candles.ValidationIssues` to highlight the missing piece(s) of data